urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/driver/INDINumberElement.java

```diff
@@ -0,0 +1,450 @@
+   */
+  private String getNumberAsString(double number) {
+    String aux;
+
+    if (numberFormat.endsWith("m")) {
+      aux = sFormatter.format(number);
+    } else {
+      Formatter formatter = new Formatter(Locale.US);
+      aux = formatter.format(getNumberFormat(), number).toString();
+    }
+
+    return aux;
+  }
+
+  @Override
+  public Double getValue() {
+    return value;
+  }
+
+  @Override
+  public void setValue(Object newValue) throws IllegalArgumentException {
+    if (newValue instanceof String) {
+      setValueAsString((String)newValue);
+    } else if (newValue instanceof Double) {
+      setValueAsdouble(((Double)newValue).doubleValue());
+    } else {
+      throw new IllegalArgumentException("Value for a Number Element must be a String or a Double");
+    }
+  }
+
+  /**
+   * Sets the maximum of this Number Element
+   *
+   * @param maxS A String with the maximum value of this Number Element
+   */
+  private void setMax(String maxS) {
+    max = parseNumber(maxS);
+  }
+
+  /**
+   * Sets the minimum of this Number Element
+   *
+   * @param maxS A String with the minimum value of this Number Element
+   */
+  private void setMin(String minS) {
+    min = parseNumber(minS);
+  }
+
+  /**
+   * Sets the step of this Number Element
+   *
+   * @param maxS A String with the step value of this Number Element
+   */
+  private void setStep(String stepS) {
+    step = parseNumber(stepS);
+  }
+
+  /**
+   * A convenience method to set the value represented by a String.
+   * @param valueS The value
+   * @throws IllegalArgumentException if it is not a value within the limits.
+   */
+  private void setValueAsString(String valueS) throws IllegalArgumentException {
+    value = parseNumber(valueS);
+
+    if ((value < min) || (value > max)) {
+      throw new IllegalArgumentException(getName() + " ; " + "Number (" + valueS + ") not in range [" + min + ", " + max + "]");
+    }
+  }
+
+  /**
+   * A covenience method to set the value represented by a double
+   * @param value The value
+   * @throws IllegalArgumentException if it is not a value within the limits.
+   */
+  private void setValueAsdouble(double value) throws IllegalArgumentException {
+    this.value = value;
+
+    if ((value < min) || (value > max)) {
+      throw new IllegalArgumentException(getName() + " ; " + "Number (" + value + ") not in range [" + min + ", " + max + "]");
+    }
+  }
+
+  /**
+   * Parses a number according to the Number Format of this Number Element.
+   *
+   * @param number The number to be parsed.
+   * @return the parsed number
+   * @throw IllegalArgumentException if the <code>number</code> is not correctly
+   * formatted.
+   */
+  private double parseNumber(String number) throws IllegalArgumentException {
+    double res;
+
+    if (numberFormat.endsWith("m")) {
+      res = sFormatter.parseSexagesimal(number);
+    } else {
+      try {
+        res = Double.parseDouble(number);
+      } catch (NumberFormatException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)